### PR TITLE
doc: introduce Code Review Process documentation

### PR DESCRIPTION
**Pull request by OddBloke**

### CODE REVIEW

**Comment by OddBloke** [comment]
The process document itself is ready for review, but the PR as a whole is not ready to land until I've updated the HACKING doc to reference it, so I've marked this "wip".

File: doc/rtd/topics/code_review.rst [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+(CI) testing in Travis is passing.  Once a PR is passing CI, it moves
+into the Review phase.
+
+These are the steps that comprise the opening phase:
+
+1. The **Proposer** opens PR
+
+2. An automated comment will be added to the PR, outlining the steps
+   expected of the **Proposer** before a **Committer** will review
+   their PR.  (N.B. This automation is not yet implemented!)
```
Comment (powersj) [inline doc/rtd/topics/code_review.rst]: Will you remember to pull this comment out once done? :)

File: doc/rtd/topics/code_review.rst [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+     assigned reviewer), effectively moving back to the start of this
+     section.
+
+Inactive Pull Requests
+======================
+
+PRs may be closed if they have been waiting on **Proposer** action for
+a certain amount of time without activity.  A PR will be marked as
+stale (with an explanatory comment) after 14 days of inactivity.  It
+will be closed after a further 7 days of inactivity.
```
Comment (powersj) [inline doc/rtd/topics/code_review.rst]: Do we have a way to avoid closing when we are waiting on the reviewer?

e.g. My PR needed fixes, I fixed them and @you to get your attention, but you didn't take a look after 14 days.

File: doc/rtd/topics/code_review.rst (hunk 1 above)
Comment (OddBloke) [reply doc/rtd/topics/code_review.rst]: We'll find out! I figured that including the step with the proviso was better than excluding it until we implemented it, but I'm happy to pull it out if you think that's better?

Comment (powersj) [reply doc/rtd/topics/code_review.rst]: Nah let's leave it for now. But I will need to give you a hard time if/when we forget :P

File: doc/rtd/topics/code_review.rst (hunk 2 above)
Comment (OddBloke) [reply doc/rtd/topics/code_review.rst]: I don't think we have a robust way of doing it presently, no. A later iteration of this, with automation, might have labels that we could use to disable the messaging.

Perhaps we should include something like "(If this PR has been waiting on the _reviewer_ for 14 days, please ping @powersj who will apologise and ensure someone looks at it.)" in the automated messaging, as a backstop?

Comment (powersj) [reply doc/rtd/topics/code_review.rst]: :+1: I really like that idea of escalating to me

File: doc/rtd/topics/code_review.rst [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+
+   (If the **Proposer** is a core developer, then they are not
+   considered part of this group for the purposes of this process.)
+
+Opening Phase
+=============
+
+In this phase, the **Proposer** is responsible for all actions. They
+should get a PR into good enough shape that it is worth **the
+Committers** spending time reviewing it.  Specifically, they are
```
Comment (igalic) [inline doc/rtd/topics/code_review.rst]: if it's not in good enough shape yet, the proposer could open the pr as WIP, or draft

File: doc/rtd/topics/code_review.rst [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+
+     The review process is **DONE**.
+
+   Approve (with nits)
+     If the **Proposer** submits their PR with "Allow edits from
+     maintainer" enabled, and the only changes the **Committer**
+     requests are minor "nits", the **Committer** can push fixes for
+     those nits and squash merge.  (If a **Proposer** wants to opt-out
+     of this, then they should uncheck "Allow edits from maintainer"
+     when submitting their PR.)
```
Comment (igalic) [inline doc/rtd/topics/code_review.rst]: there's also the option to propose a change in github terms

Comment (OddBloke) [reply doc/rtd/topics/code_review.rst]: This feels like a tool that a reviewer can use, but I don't know that it really changes the code review _process_. The pull request will still either go via "Approve with nits" or "Request changes", "propose a change" is just a different way of expressing the nits and/or requested changes for those paths.

If you'd still like to see something added or change, could you propose a specific change that we can use as a starting point for discussion?

(Thanks for the review!)

File: doc/rtd/topics/code_review.rst (hunk 3 above)
Comment (OddBloke) [reply doc/rtd/topics/code_review.rst]: Hmm, this is an interesting point. We need to strike a balance here: we don't want to encourage people to submit half-baked PRs in the hopes that we will pick it up and do the work for them, but we also do want people (like you) who are working on chunkier changes to not have to wait until they have a full change to propose for us to get eyes on it.

My gut feeling is that the people in the latter camp will all be involved in development on a regular basis, so we can discuss exceptions to the process in IRC or bug comments. This perhaps means that we don't need to capture this into the process itself? What do you think?

(Also, as a side note: not everyone can apply/remove labels, so we can't use them as a self-serve process step. I had to revise an earlier draft of this process once I realised that.)

File: doc/rtd/topics/code_review.rst (hunk 2 above)
Comment (OddBloke) [reply doc/rtd/topics/code_review.rst]: I've added this to the relevant PR: https://github.com/canonical/cloud-init/pull/125/files#diff-12eea56dbf5e1eb25defd0d34cc314bbR21

File: doc/rtd/topics/code_review.rst [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+Proposer
+   The developer proposing a pull request (hereafter known as a PR).
+   This could be a core developer or community member.  (This is likely
+   the role you will be taking in this process.)
+
+the Committers
+   The group of cloud-init core developers.
+
+   In this process doc, cloud-init core developers are treated as a
+   bloc of interchangeable people.  This is a simplification, as in
```
Comment (rjschwei) [inline doc/rtd/topics/code_review.rst]: s/bloc/block/

File: doc/rtd/topics/code_review.rst [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+   reality we will need to manage transfer of reviews, dismissal of
+   stale reviews, parallel reviews from multiple committers, etc.  If
+   we find that handling these on an ad-hoc basis is causing problems,
+   then we should revisit and expand this process to include them.
+
+   (If the **Proposer** is a core developer, then they are not
+   considered part of this group for the purposes of this process.)
+
```
Comment (rjschwei) [inline doc/rtd/topics/code_review.rst]: The way this is structure unfortunately discourages others from reviewing code, IMHO. The core team should have keen interest in training and enabling additional core team members. Doing code reviews is one of the primary training tools for such growth. If code reviews by others are more or less actively discouraged by way of the implications as written here it is going to make things more difficult.

Over all I would suggest that we switch the use of "Committer" to "Reviewer" in the context of reviews. This then allows for clearer separation of the tasks of review and "merge/acceptance" in that the group "Reviewer" and "Committer" get separate definitions. Effectively we would have 3 role definitions in this section.

At some point during the process a "Committer" takes on the role of "Reviewer" but really anyone, spare the "Proposer" should be able to assume the role of "Reviewer" and that voice should be heard. Obviously a review by a core team member is required for the merge. That can be stated in a way that does not discourage others from being "Reviewers".

File: doc/rtd/topics/code_review.rst (hunk 3 above)
Comment (rjschwei) [reply doc/rtd/topics/code_review.rst]: "good enough" has to be defined. After all early on it was stated that we have this document to remove ambiguity. The definition of what "good enough" means should occur prior to using the term.

Further there is the case where one needs advice/feedback before proceeding. Such conversations are best captured in the conversation in a PR, thus "good enough" may actually not apply. ANd to folow the finger wagging tone of the document ;) it better be worth "the committers" time to provide feedback on something that is WIP to help unblock someone that is working on code and got stuck.

Lastly this can be written without wagging the finger, which is the impression the current chosen wording provides.

Core team members, as well as all other contributors should be here to help. If that help is to provide feedback to someone that might be new to the project and whose initial attempt may not be what one would ordinarily consider to be in "good shape" then IMHO it is still worth while to help that person along. The wording here doesn't really make me feel welcome and encouraged to contribute to the project.

I think it is worth to define various categories here for example:
- Needs feedback on approach
- Needs help, I am stuck
- Ready for merge review (and one criteria for this would be all tests pass)

Then we can define the process of adding this categorization into the comments since not everyone can set labels. Something like this also clarifies the state of the PR. For example "Needs help, I am stuck" should have a very specific question such as "this test is failing and I don't know why. I've looked here and there but cannot figure it out" or "I do not understand the structure of the test setup in XYZ and thus cannot add tests for my new code here, can you please explain"

I think this section needs more clarity and differentiation and yes the wording should aim to be inclusive and friendly.

File: doc/rtd/topics/code_review.rst [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+into the :ref:`Review phase`.
+
+These are the steps that comprise the opening phase:
+
+1. The **Proposer** opens PR
+
+2. An automated comment will be added to the PR, outlining the steps
+   expected of the **Proposer** before a **Committer** will review
+   their PR.  (N.B. This automation is not yet implemented!)
+
```
Comment (rjschwei) [inline doc/rtd/topics/code_review.rst]: As stated above I think there are different classes of "requests for reviews" I think those will need to be taken into account in any automation.

File: doc/rtd/topics/code_review.rst [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+      assistance.  Alternatively, for more immediate assistance, they
+      can ask in ``#cloud-init`` on the Freenode IRC network.
+
+   CI passes
+      Move on to the :ref:`Review phase`.
+
+Review Phase
+============
+
+In this phase, the **Proposer** and the **Committers** will iterate
```
Comment (rjschwei) [inline doc/rtd/topics/code_review.rst]: This is another place where the wording more or less actively discourages people that are not in the core team from reviewing the code. Yes the core team members have ultimate say w.r.t. what gets merged but as stated earlier it shold be the desire of the core team to grow and have more members to reduce the review work load over all.

File: doc/rtd/topics/code_review.rst [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+There are three potential outcomes: merged, rejected permanently, and
+temporarily closed.
+
+These are the steps that comprise the review phase:
+
+1. **The Committers** assign a **Committer** to the PR
+
+   The cloud-init core developers will work amongst themselves to
+   ensure that this happens in a timely fashion.  (Generally,
+   assignment will happen at the team's daily internal meeting.)  Note
```
Comment (rjschwei) [inline doc/rtd/topics/code_review.rst]: Having "internal meetings" for an open source project is a bad idea. The meeting should be public, we have IRC for this it should be used.

File: doc/rtd/topics/code_review.rst [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+   that, in GitHub terms, this is setting an Assignee, not requesting a
+   review.
+
+2. That **Committer** reviews the PR, resulting in one of the
+   following:
+
+   Approve
+     The **Committer** will squash merge immediately.
```
Comment (rjschwei) [inline doc/rtd/topics/code_review.rst]: Hmm maybe the squashing should be the proposers responsibility and one of he requirements for the "Ready for merge review" state is that the RR has one commit.

File: doc/rtd/topics/code_review.rst [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+
+     There may be circumstances where a PR should not be merged
+     immediately; these will be handled case-by-case, and the
+     **Proposer** should make this requirement very clear in their pull
+     request description.
+
```
Comment (rjschwei) [inline doc/rtd/topics/code_review.rst]: Well there may also be the case where a core team member knows of something in flight and the PR should be held until after the in flight work is complete. This case should be documented here as well.

File: doc/rtd/topics/code_review.rst (hunk 4 above)
Comment (rjschwei) [reply doc/rtd/topics/code_review.rst]: "Approve with nits" is basically a conditional approval. These IMHO are not helpful. The PR is either approved and can be merged or it needs changes. This of course does not resolve the cases where there are nits, which by definition are more or less things that can be debated forever without a real answer or satisfactory conclusion.

So if the reviewer has nits the reviewer should open a PR against the branch of the proposer and then the proposer can merge the work and the PR then moves forward. If the reviewer has nits but hey are not worth the reviewers time to submit a PR then they are not worth fixing and the PR should be merged. 

My $0.02

File: doc/rtd/topics/code_review.rst (hunk 10 above)
Comment (OddBloke) [reply doc/rtd/topics/code_review.rst]: We squash merge through the GitHub UI, let me make that explicit.

File: doc/rtd/topics/code_review.rst (hunk 5 above)
Comment (OddBloke) [reply doc/rtd/topics/code_review.rst]: Per https://www.dictionary.com/browse/bloc:

> a group of persons, businesses, etc., united for a particular purpose.

:-)

File: doc/rtd/topics/code_review.rst (hunk 6 above)
Comment (OddBloke) [reply doc/rtd/topics/code_review.rst]: That's a great point, thank you!

File: doc/rtd/topics/code_review.rst (hunk 9 above)
Comment (OddBloke) [reply doc/rtd/topics/code_review.rst]: Broadly speaking, I agree. The internal meetings referred to here are the Canonical team's daily standup meetings, which happen via video call, and are our only regular daily touchpoint. `cloud-init` is only one of the projects that are covered by our team, and the other projects are not all open source.

If we want to reliably do this daily then, practically, we're going to be doing it at our internal meeting.

File: doc/rtd/topics/code_review.rst (hunk 11 above)
Comment (OddBloke) [reply doc/rtd/topics/code_review.rst]: Agreed.

File: doc/rtd/topics/code_review.rst (hunk 4 above)
Comment (OddBloke) [reply doc/rtd/topics/code_review.rst]: OK, I'll need to reword this to state the overall outcome first, and then the details, because it specifically _isn't_ a conditional approval. This is the crucial part (bold added for this comment):

> the Committer can push fixes for those nits and **squash merge**.

"Approve (with nits)" is a committer saying, "this is an Approve except for these nits, let me fix those for you (to save a review round-trip) and merge".

Reading the current text again with fresh eyes, this could be a lot clearer.

File: doc/rtd/topics/code_review.rst (hunk 9 above)
Comment (rjschwei) [reply doc/rtd/topics/code_review.rst]: Well, but that creates another barrier to entry to anyone that might aspire to become a core team member/committer and that does not happen to work at Canonical. While we do not have such a person it really makes not difference. But at the very least the wording chosen should reflect that possibility. I'd suggest something along these lines:

"""
PR reviews for final approval are picked up by core team members or assigned to core team members in a daily meeting.
"""

Not being specific about the "location" of the meeting allows for the possibility that an external to Canonical core team member can exist and can than participate in that meeting.

**Comment by OddBloke** [comment]
@rjschwei Thank you for the thorough review. I've responded to some of your comments inline when I had something specific to say, and will look at the others more deeply after the weekend. Broadly speaking, I think I've focused too much on what we (as committers) need to do to make the review process work better, and then filled in the gaps of what the community should be doing (which is why those parts are not up to scratch). And that focus is also why I've drawn such a distinction between committers and community members ("what can _we_ do to help _them_"), which is also not helpful. Your outside perspective is much appreciated, thank you again!

File: doc/rtd/topics/code_review.rst (hunk 9 above)
Comment (OddBloke) [reply doc/rtd/topics/code_review.rst]: Good advice, thank you!

**Comment by OddBloke** [comment]
OK, I've been giving this some thought over the past couple of days (in between suffering from a cold, hence the delay!). I'm going to outline my intended changes, to get feedback before I start actually making them.

My plan is to revise the document to have _three_ roles as suggested by @rjschwei: Proposer, Reviewer and Committer. The process will be reworked to focus on the interaction between Proposer and Reviewer, with Committer simply being a special-case of a Reviewer. (A review from a Committer will still be required to land changes, of course.) I will also integrate the inline suggestions.

I agree that this draft comes off as condescending, in places and as a whole. I attempted to lighten the tone of the initial draft right before submitting it, and I don't think that worked very well. I'm hopeful that the improved definition of roles will cause a lot of that to fall out naturally, but I will also be mindful of that when drafting the next round.

Once again, thank you to everyone who's taken a look! Redrafting this is top of my list, so if you have any concerns, questions or comments about the above, please let me know!

**Comment by OddBloke** [comment]
(When I wrote that last comment, I'd forgotten I was taking tomorrow off. I'll be looking at this again next week!)

**Comment by blackboxsw** [comment]
@OddBloke I left this unmerged because it still has a `wip` label

**Comment by OddBloke** [comment]
@rjschwei This is ready for another pass from you, if you have time.

File: doc/rtd/topics/code_review.rst [hunk 12]
```diff
@@ -0,0 +1,254 @@
...
+   A cloud-init core developer (i.e. a person who has permission to
+   merge PRs into master).
+
+Prerequisites For Landing Pull Requests
+=======================================
+
+Before a PR can be landed into master, the following conditions *must*
+be met:
+
+* the CLA has been signed by the **Proposer**
```
Comment (rjschwei) [inline doc/rtd/topics/code_review.rst]: "or entity agreement exists" should be added

Comment (OddBloke) [reply doc/rtd/topics/code_review.rst]: Done.

**Comment by OddBloke** [comment]
@rjschwei I've tried to integrate the inactive PRs section in a little better in this latest iteration.

**Comment by rjschwei** [comment]
@OddBloke LGTM thanks
